### PR TITLE
Add some docs about architectural design

### PR DESCRIPTION
**Pull request by mdellweg**
I started to draft this and then i realized, there might be some overlap with what is in README.md.

Should we put all that stuff in README, or is it time to think about generating proper docs with navigating and searching?
(I sure hope that the _use_ of the cli is mainly selfexplanatory.)

### CODE REVIEW

**Comment by daviddavis** [comment]
Is this ready for review?

**Comment by daviddavis** [comment]
> I started to draft this and then i realized, there might be some overlap with what is in README.md.

> Should we put all that stuff in README, or is it time to think about generating proper docs with navigating and searching?

I think overlap is fine. I like having stuff separate from the README because I think it'll grow. We should think about publishing the docs I think at some point.

**Comment by mdellweg** [comment]
Bring on the review!

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Pulp CLI Architecture
+
+## Plugin system
+
+The Pulp CLI is designed with a plugin structure. Plugins can either live in the main package or be shipped independently.
```
Comment (daviddavis) [inline docs/architecture.md:5]:
```suggestion
The Pulp CLI is designed with a plugin structure. Plugins can either live in the pulp-cli package or be shipped independently.
```

Comment (daviddavis) [reply docs/architecture.md:5]: I'm assuming that `main` is the module name and pulp-cli is the name of the package. Or maybe I misunderstand?

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from pulpcore.cli.common import main
+
+@main.command()
+def my_command():
+    pass
+```
+
+## Contexts
+
+In `click`, every subcommand is accompanied by a `click.Context`, and objects can be attached to them.
```
Comment (daviddavis) [inline docs/architecture.md:29]: Is this always true? If not, I'd suggest:

```suggestion
In `click`, every subcommand is usually accompanied by a `click.Context`, and objects can be attached to them.
```

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+In this CLI we attach a `PulpContext` to the main command.
+It abstracts the communication to the pulp server through its `api` property.
```
Comment (daviddavis) [inline docs/architecture.md:31]:
```suggestion
It handles the communication to the pulp server through its `api` property.
```

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+Further we encourage to abstract the communication with certain endpoints by subclassing the `PulpEntityContext`.
```
Comment (daviddavis) [inline docs/architecture.md:32]:
```suggestion
Further we encourage the handling of communication with certain endpoints by subclassing the `PulpEntityContext`.
```

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+@pass_entity_context
+def my_sub_command(entity_ctx):
+    ... href = ...
+    entity_ctx.destroy(href)
+```
+
+### Version dependent codepaths
+
+The pulp CLI is designed to support a number of pulp server versions at the same time.
+The entity contexts should be used to implement necessary server version depedent workarounds.
```
Comment (daviddavis) [inline docs/architecture.md:54]:
```suggestion
The entity contexts should be used to implement necessary server version-dependent workarounds.
```

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+@my_command.command()
+@pass_entity_context
+def my_sub_command(entity_ctx):
+    ... href = ...
+    entity_ctx.destroy(href)
+```
+
+### Version dependent codepaths
+
+The pulp CLI is designed to support a number of pulp server versions at the same time.
```
Comment (daviddavis) [inline docs/architecture.md:53]:
```suggestion
Each pulp CLI release is designed to support multiple pulp server versions.
```

File: docs/architecture.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+@pass_pulp_context
+@click.pass_context
+def my_command(ctx, pulp_ctx):
+    pulp_ctx.needs_plugin("pulp_my_content", min_version="1.0.0")
+    ctx.obj = MyEntityContext(pulp_ctx)
+```
+
+## Generics
+
+For certain regularly repeating patterns, like listing all entities of a kind, we provide generic commands that will use the underlying context objects.
```
Comment (daviddavis) [inline docs/architecture.md:76]:
```suggestion
For certain often repeated patterns like listing all entities of a  particular kind, we provide generic commands that use the underlying context objects.
```

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+```python
+from pulpcore.cli.common.generic import name_option, show_command,
+
+lookup_params = [name_option]
+my_command.add_command(show_command(decorators=lookup_params))
+```
+
+Some of these commands understand extra arguments.
+They can be attached by passing a list of `click.Options` via the decorators argument.
```
Comment (daviddavis) [inline docs/architecture.md:86]:
```suggestion
They can be attached by passing a list of `click.Options` via the `decorators` argument.
```

**Review by daviddavis — APPROVED** [review]
Added some suggestions but nothing blocking.

File: docs/architecture.md (hunk 1 above)
Comment (mdellweg) [reply docs/architecture.md:5]: `main` is only the command entrypoint to be precise. And precise is good!

File: docs/architecture.md (hunk 2 above)
Comment (mdellweg) [reply docs/architecture.md:29]: It is. The command you are writing and decorating is only called via `ctx.invoke()`.
Optional is the use of `ctx.obj`.

File: docs/architecture.md (hunk 3 above)
Comment (mdellweg) [reply docs/architecture.md:31]: What about?
```suggestion
It acts as an abstraction layer that handles the communication to the pulp server through its `api` property.
```
Would this imply that we do not accept contributions that disregard the abstraction and call e.g. requests directly?

File: docs/architecture.md (hunk 5 above)
Comment (mdellweg) [reply docs/architecture.md:54]: trust a native speaker...

**Pull request merged**